airflow/models.py

```diff
@@ -999,27 +999,31 @@ def are_dependencies_met(
         task = self.task
 
         # Checking that the depends_on_past is fulfilled
-        if (task.depends_on_past and not ignore_depends_on_past and
-                not self.execution_date == task.start_date):
-            previous_ti = session.query(TI).filter(
-                TI.dag_id == self.dag_id,
-                TI.task_id == task.task_id,
-                TI.execution_date ==
-                    self.task.dag.previous_schedule(self.execution_date),
-                TI.state.in_({State.SUCCESS, State.SKIPPED}),
-            ).first()
-            if not previous_ti:
-                if verbose:
-                    logging.warning("depends_on_past not satisfied")
-                return False
-
-            # Applying wait_for_downstream
-            previous_ti.task = self.task
-            if task.wait_for_downstream and not \
-                    previous_ti.are_dependents_done(session=session):
-                if verbose:
-                    logging.warning("wait_for_downstream not satisfied")
-                return False
+        # 1. get the last TI run prior to the current execution date
+        # 2. if there is one and it didn't succeed/skip, the condition fails
+        if (task.depends_on_past and not ignore_depends_on_past):
+            previous_ti = (
+                session.query(TI)
+                .filter(
+                    TI.dag_id == self.dag_id,
+                    TI.task_id == task.task_id,
+                    TI.execution_date < self.execution_date)
```

Comment:
I don't think the index on the table are covering this. You might want to add an index on dag_id,task_id,execution_date